lms/services/assignment.py

```diff
@@ -10,55 +15,206 @@ def __init__(self, db):
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                ext_lti_assignment_id=ext_lti_assignment_id,
+            )
+            .one()
+        )
+
+    def _get_for_canvas_launch(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id,
+        ext_lti_assignment_id,
+    ):
+        """
+        Get a canvas assignment by both resource_link_id and ext_lti_assignment_id.
+
+        Having both values means that we are most likely launching an assignment but
+        we might find these cases on the DB data:
+
+            - A row with both values.
+                An assignment that has been launched before.
+                Just get it an return it.
+
+            - A row with with just the `ext_lti_assignment_id`.
+                We have configured this assignment before but never launched it
+                so we don't yet know its `resource_link_id`.
+
+                Get it, fill the new value on `resource_link_id` and return it.
+
+            - Two rows, one with each of the ids an the other set to null.
+
+                In this case we stored the assignment in the past on launch with just the `resource_link_id`,
+                later, we re-configured it, creating a new row with just the `ext_lti_assignment_id`.
+
+                We need to get both and merge them into one.
+                We'll keep the one with `ext_lti_assignment_id` as is the newest one.
+        """
+        query = self._db.query(Assignment).filter(
+            Assignment.tool_consumer_instance_guid == tool_consumer_instance_guid,
+            (
+                # Regular canvas launch
+                (
+                    (Assignment.resource_link_id == resource_link_id)
+                    & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                )
+                # Configured file assignment that was stored in the DB
+                # before all canvas assignments were stored in the DB
+                | (
+                    (Assignment.resource_link_id == resource_link_id)
+                    & (Assignment.ext_lti_assignment_id.is_(None))
+                )
+                #  First launch of a newly configured assignment
+                | (
+                    (Assignment.resource_link_id.is_(None))
+                    & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                )
+            ),
+        )
+
+        assignments = query.all()
+        if not assignments:
+            raise NoResultFound()
+
+        if len(assignments) == 2:
+            assignment = self._merge_canvas_assignments(assignments)
+        else:
+            assignment = assignments[0]
+
+        if resource_link_id and not assignment.resource_link_id:
+            # First lunch of a canvas assignment, fill the resource_link_id now
+            assignment.resource_link_id = resource_link_id
+
+        return assignment
```

Comment:
I think it might be good to use this Python code where each of the different cases is dealt with as the place to put explanatory comments, rather than in the docstring or the SQLAlchemy query above. Putting the explanations of the cases next to the code that handles the cases gives the comments a better chance of being kept up to date. Something like this:

```python
if len(assignments) == 2:
    # We found two assignments: one with the matching resource_link_id and no ext_lti_assignment_id
    # and one with the matching ext_lti_assignment_id and no resource_link_id.
    #
    # This happens because legacy code used to store Canvas assignments in the DB with a
    # resource_link_id and no ext_lti_assignment_id, see https://github.com/hypothesis/lms/pull/2780
    #
    # Whereas current code stores Canvas assignments during content-item-selection with an
    # ext_lti_assignment_id and no resource_link_id.
    #
    # We need to merge the two assignments into one.
    assignment = self._merge_canvas_assignments(assignments)
else:
    assignment = assignments[0]

if resource_link_id and not assignment.resource_link_id:
    # We found an assignment with an ext_lti_assignment_id but no resource_link_id.
    # This happens the first time a new Canvas assignment is launched: the assignment got created
    # during content-item-selection with an ext_lti_assignment_id but no resource_link_id,
    # and then the first time the assignment is launched we add the resource_link_id.
    assignment.resource_link_id = resource_link_id

# We found an assignment with the matching resource_link_id and ext_lti_assignment_id.
return assignment
```

Those comments are pretty long but I think this might be an appropriate time for some long comments. The edge cases we're covering here are pretty damn obscure, especially the merge-assignments one. I'm not sure the wording of my suggested comments is ideal though, could probably be improved.